app/scripts/modules/core/account/account.service.ts

```diff
@@ -0,0 +1,210 @@
+              private cloudProviderRegistry: any,
+              private API: Api) {
+  }
+
+  public challengeDestructiveActions(account: string): ng.IPromise<any> {
+    return memoize(() => {
+      return this.$q((resolve: ng.IQResolveReject<any>) => {
+        if (account) {
+          this.getAccountDetails(account)
+            .then((details: any) => {
+              if (details) {
+                resolve(details.challengeDestructiveActions);
+              } else {
+                resolve(false);
+              }
+            })
+            .catch(() => resolve(false));
+        } else {
+          resolve(false);
+        }
+      });
+    })();
+  }
+
+  public getAccountDetails(account: string): any {
+    return this.API.one('credentials')
+      .one(account)
+      .useCache()
+      .get();
+  }
+
+  public getAllAccountDetailsForProvider(provider: string): ng.IPromise<any> {
+    return memoize(() => {
+      return this.listAccounts(provider)
+        .then((accounts: IAccount[]) => this.$q.all(accounts.map((account: IAccount) => this.getAccountDetails(account.name))))
+        .catch((error: any) => {
+          this.$log.warn(`Failed to load accounts for provider "${provider}"; exception:`, error);
+          return [];
+        });
+    })();
+  }
+
+  public getAvailabilityZonesForAccountAndRegion(provider: string, account: string, region: string): ng.IPromise<any[]> {
+    return this.getPreferredZonesByAccount(provider)
+      .then((result: any) => result[account] ? result[account][region] : []);
+  }
+
+  public getCredentialsKeyedByAccount(provider: string): ng.IPromise<any> {
+    return memoize(() => {
+      return this.$q((resolve: ng.IQResolveReject<any>) => {
+        this.listAccounts(provider)
+          .then((accounts: IAccount[]) => {
+            this.$q.all(accounts.reduce((aggregated: any, account: IAccount) => {
+              aggregated[account.name] = this.API.one('credentials').one(account.name).useCache().get();
+              return aggregated;
+            }, {}))
+              .then((result: any) => resolve(result));
+          });
+      });
+    })();
+  }
+
+  public getPreferredZonesByAccount(provider: string): ng.IPromise<IAccountZone> {
+    return memoize(() => {
+      const preferred: IAccountZone = {};
+      return this.getAllAccountDetailsForProvider(provider)
+        .then((accounts: IAccount[]) => {
+          accounts.forEach((account: IAccount) => {
+            preferred[account.name] = {};
+            account.regions.forEach((region: IRegion) => {
+              let zones: string[] = region.availablityZones;
+              if (region.preferredZones) {
+                zones = intersection(region.preferredZones, region.availablityZones);
+              }
+
+              preferred[account.name][region.name] = zones;
+            });
+          });
+
+          return preferred;
+        });
+    })();
+  }
+
+  public getRegionsForAccount(account: string): any[] {
```

Comment:
an `IRegion` interface would be nice, but, again, not sure how standard it's going to be - might just have `name`, based on what I'm seeing from AWS and Titus credentials.